### PR TITLE
Core/Movement: Properly calculate the first Catmull-Rom spline point

### PR DESCRIPTION
**Pull request by DanVS**
**Changes proposed:**
Implement #22435.
Credit @xvwyh.

**Target branch(es):** 3.3.5

**Tests performed:** _Somewhat_ tested in-game.

### CODE REVIEW

**Comment by Kittnz** [comment]
Any tests preformed?

**Comment by DanVS** [comment]
Well, seems to be working, but right now I am unable to perform tests with multiple accounts, so this probably requires attention.

**Comment by DanVS** [comment]
Works on flying mobs from Icecrown and The Storm Peaks.

File: src/server/game/Movement/Spline/Spline.cpp [hunk 1]
```diff
@@ -253,14 +255,14 @@ void SplineBase::InitCatmullRom(Vector3 const* controls, index_type count, index
         if (cyclic_point == 0)
             points[0] = controls[count-1];
         else
-            points[0] = controls[0].lerp(controls[1], -1);
+            points[0] = controls[0] - G3D::Vector3{ cos(initialOrientation), sin(initialOrientation), 0.0f };
```
Comment (Shauren) [inline src/server/game/Movement/Spline/Spline.cpp]: use std::sin and std::cos (also in the other line)

**Review by Shauren — CHANGES_REQUESTED** [review]

File: src/server/game/Movement/Spline/Spline.cpp (hunk 1 above)
Comment (DanVS) [reply src/server/game/Movement/Spline/Spline.cpp]: Fixed. Thanks!

**Review by Shauren — APPROVED** [review]

**Comment by Ovahlord** [comment]
@xvwyh summoning you here for a semi related discussion. I'm currently digging into splines again and I have noticed something while prototyping smooth waypoint transitions and that is that creatures tend to have some small setback hickups when stopping or launching new splines while another spline is going on. 

So far I've figured out that's simply because the clientside position is ahead of the serverside one which makes sense considering there are things like latency so how does Blizz actually prevent these hickups? Do they have some 'lookahead value' internally which they use to offset splines so they compute a position that's somewhat more sync with the client?

Or do they send splines with a slightly increased duration instead so the unit moves a tad slower to mitigate that?

**Comment by xvwyh** [comment]
What do you mean by "setback hickups"? Also what spline type are we talking about: linear or catmull-rom?

I've dug into the client and found the code that makes the creature "backtrack" to the first point of a linear spline if creature's clientside position differs too much from the first path point. E.g. if we have the following situation:
```
<----C----S-----




          X
```
(where C is clientside's position of a creature that's currently moving left and is ahead of the server for whatever reason, S is its serverside's position and X is the destination of a new movement)
You might think that the client will plot the new spline like this, matching the path that the server sent to the client:
```
     C    S     
          |
          |
          |
          V
          X
```
but it won't, because that will cause the creature to teleport abruptly to it's serverside's starting position.

So then you might think that the client will plot the new spline like this:
```
     C    S     
      \
       \
        \
         V
          X
```
but it won't, because that will cause severe discrepancy in the creature's positions throughout that path from that of the server (the server will send the creature straight downwards in this situation)

What happens instead is the creature's current clientside position is added to the front of the path:
```
     C--->S     
          |
          |
          |
          V
          X
```
so the creature will backtrack to where it's supposed to be from server's PoV, and then proceed with the intended path. This happens without changing the duration of the spline! So yes, it will be sped up, and it won't precisely match server's positions throughout the movement, but at least the path taken will be roughly the same, and the creature will arrive in its destination at the same time as on the server.

The pseudocode for it is roughly
```cpp
if (GetDistanceBetween(GetClientsidePosition(), path.front()) >= 0.027777777)
    path.push_front(GetClientsidePosition());
else
    path.clear(); // This seems weird? Keep reading, I made a mistake, but I'm not rewriting this wall of text
```
image

(`points` here is a pointer pointing to the SECOND element of an array of 3d vectors, the FIRST element is initially unused and reserved specifically for this case: it gets filled by the unit's current clientside position, count of points is incremented, and then the pointer to the first element is returned, allowing it to be used in the spline)

But. There's always a but. This works only when the path consists of just a start and destination points. No other intermediate points.

When the path has more points, some **very** convoluted code is executed, that searches for some specific point in that path. There's a similar code in alpha client, and it's way more straightforward in there - it simply searches for the point that's closest to unit's current position. The code in WotLK client is much more complicated, and it seems to even have some loop unrolling in it, so it's a nightmare to read, and I couldn't figure out what it was doing.  But at the end of it, depending on the index of the point that was found, one of several cases can happen:
a) Same thing happens as written above: current clientside position gets added to the front of the path (it checks if 2D (!) distance between the current position and the found point is >= 1)
b) The whole path is replaced with just 2 points: current clientside position and final destination (it checks if 3D distance between the current position and the found point is >= 0.027777777)
c) The whole path is erased and no movement happens
d) Even more convoluted code is executed that searches the path for yet another point, checking sometimes 2D sometimes 3D distance and then the path gets cut at that point and either the former or latter part of the path is discarded and jesus fucking christ I'm done, I'm out, I don't want to see this ever again.

Actually, no, I lied, the case I described with only start and destination points should actually be the case when there is NO movement at all, or the distance between start and dest points is < 0.027777777, in which case no movement happens either. But, yeah, if the clientside position is too far from it - the unit WILL move back to where it should be based on the data from the packet, and now that `path.clear()` make sense. But I'm too lazy to rewrite everything I wrote already, because I doubt it will help you anyway.

**Comment by xvwyh** [comment]
And what do you mean by "smooth waypoint transitions"? Did you mean simply that you want to prevent the unit from stopping in each point for a short time awaiting a new movement packet from the server? In that case just send the whole path in one single movement packet and you're done...

**Comment by Ovahlord** [comment]
Well, as far as I'm concerned, it happens for all splines. Imagine a creature chases you and you you kill that creature during chasing, there is a very high chance that the unit will turn arround and look towards its opposite direction because that's where the server says where the creature is suposed to be.

I went ahead and did some packet research and the first thing that I noticed is that the spline duration in retail packets slightly differs from what you get when doing manual calculations (something between 1 - 80ms differences).

However, I did further experiments by computing positions ahead of time which solves these turnarrounds but causes ugly snapping. At first I just added the mstime between the current time and the last gametime update cycle which makes the spline positions pin point accurate in most scenarios but then again there is this latency factor. The longer the packet takes to arrive at the client, the higher the chance that this turnarround happens again but knowing retails laggy nature there must be some trick to avoid that

**Comment by xvwyh** [comment]
The thing depicted on the video is creatures stopping on death, that's not exactly the same as going from spline to spline. Stop packets have heir own special handling in the code, and way less nightmarish:
image
(case 1:)
Checks whether the distance between the current clientside position and the position from the spline is >= cvar "pathDistTol" (1 yard by default). If it's smaller than the cvar, it simply teleports the unit to the position. If it's larger than the cvar - it proceeds onwards with creating a linear spline that causes the creature to walk to the position and thus turn around.

But here's the thing: I looked at a few sniffs, and whenever I saw a creature dying in them - it was sending a movement packet with facing type set to 0, aka it's NOT a stop packet, but a simple movement packet with a completely empty path, so only the starting position is present in the packet. Maybe it behaves differently, give it a try. Basically replace
https://github.com/TrinityCore/TrinityCore/blob/cc8287927b9c9f0bf20143972fd4d3c338bff688/src/server/game/Movement/Spline/MoveSplineInit.cpp#L186
with
https://github.com/TrinityCore/TrinityCore/blob/cc8287927b9c9f0bf20143972fd4d3c338bff688/src/server/game/Movement/Spline/MoveSplineInit.cpp#L140
and maybe make some more adjustments if it won't work out of the box, but I think it should.

**Comment by Ovahlord** [comment]
Yeah, I figured out already that dying creatures use a spline with facing 0. But their path is not empty. It's actually a path with a single point that's the exact same point as their starting point:

> ServerToClient: SMSG_ON_MONSTER_MOVE (0x6E17) Length: 50 ConnIdx: 2 Time: 08/01/2012 17:34:40.140 Number: 26112
> MoverGUID: Full: 0xF130C1840004707C Type: Creature Entry: 49540 (Stormwind Rat) Low: 290940
> VehicleExitVoluntary: 0
> Position: X: -8720.3125 Y: 456.77148 Z: 97.642105  << initial starting position
> (MovementMonsterSpline) Id: 13147604
> (MovementMonsterSpline) (MovementSpline) Face: 0 (Normal)
> (MovementMonsterSpline) (MovementSpline) Flags: 0 (None)
> (MovementMonsterSpline) (MovementSpline) MoveTime: 0
> (MovementMonsterSpline) (MovementSpline) PointsCount: 1
> (MovementMonsterSpline) (MovementSpline) Points: X: -8720.3125 Y: 456.77148 Z: 97.642105  << spline path destination point

Anyways, I've already replicated these packets with the exact same data but these turnarrounds are still the same. To be precise, they got even worse now and are much more consistent. Now pretty much every single dying creatures turns arround so I'm a bit at a loss here.

Little proof that I send it the same way as retail:

> ServerToClient: SMSG_ON_MONSTER_MOVE (0x6E17) Length: 48 ConnIdx: 0 EP: 127.0.0.1:60143 Time: 11/19/2020 02:32:30.138 Number: 396
> MoverGUID: Full: 0xF13001DB0000007A Type: Creature Entry: 475 (Kobold Tunneler) Low: 122
> VehicleExitVoluntary: 0
> Position: X: -9824.555 Y: 192.54974 Z: 12.730363
> (MovementMonsterSpline) Id: 256
> (MovementMonsterSpline) (MovementSpline) Face: 0 (Normal)
> (MovementMonsterSpline) (MovementSpline) Flags: 0 (None)
> (MovementMonsterSpline) (MovementSpline) MoveTime: 0
> (MovementMonsterSpline) (MovementSpline) PointsCount: 1
> (MovementMonsterSpline) (MovementSpline) Points: X: -9824.555 Y: 192.54974 Z: 12.730363

----
Edit: I solved the mystery. It's SMSG_ATTACK_STOP. It has a 3rd field that should be set to 1 when the attacker has died.

https://github.com/TrinityCore/TrinityCore/blob/3.3.5/src/server/game/Server/Packets/CombatPackets.cpp#L40
there are two mistakes that need to be fixed. The first one is that NowDead has to be 1 when the attacker has died, not the victim. (sniff confirmed). This field preserves the facing angle of the unit when set to 1. The 2nd one is that AttackStop is being sent before the unit's deathState gets updates so this field never gets set to 1.

**Comment by xvwyh** [comment]
Huh, in my sniffs there are no points, just the initial position. Must've changed between expansions, the oldest ones I checked were from 21463.

Hilarious that movement-related stuff is controlled from SMSG_ATTACK_STOP.

**Comment by Ovahlord** [comment]
The one I used was from 15595 so that's funny, ye

**Pull request merged**